base/libgit2/gitcredential.jl

```diff
@@ -0,0 +1,289 @@
+    if startswith(helper, '!')
+        cmd_str = helper[2:end]
+    elseif isabspath(first(Base.shell_split(helper)))
+        cmd_str = helper
+    else
+        cmd_str = "git credential-$helper"
+    end
+
+    GitCredentialHelper(`$(Base.shell_split(cmd_str)...)`)
+end
+
+function Base.:(==)(a::GitCredentialHelper, b::GitCredentialHelper)
+    a.cmd == b.cmd
+end
+
+function run!(helper::GitCredentialHelper, operation::AbstractString, cred::GitCredential)
+    cmd = `$(helper.cmd) $operation`
+    output, input, p = readandwrite(cmd)
+
+    # Provide the helper with the credential information we know
+    write(input, cred)
+    write(input, "\n")
+    t = @async close(input)
+
+    # Process the response from the helper
+    Base.read!(output, cred)
+    close(output)
+    wait(t)
+
+    return cred
+end
+
+function run(helper::GitCredentialHelper, operation::AbstractString, cred::GitCredential)
+    run!(helper, operation, deepcopy(cred))
+end
+
+# The available actions between using `git credential` and helpers are slightly different.
+# We will directly interact with the helpers as that way we can request credential
+# information without a prompt (helper `get` vs. git credential `fill`).
+# https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage
+
+fill!(helper::GitCredentialHelper, cred::GitCredential) = run!(helper, "get", cred)
+approve(helper::GitCredentialHelper, cred::GitCredential) = run(helper, "store", cred)
+reject(helper::GitCredentialHelper, cred::GitCredential) = run(helper, "erase", cred)
+
+"""
+    credential_helpers(config, git_cred) -> Vector{GitCredentialHelper}
+
+Return all of the `GitCredentialHelper`s found within the provided `config` which are valid
+for the specified `git_cred`.
+"""
+function credential_helpers(cfg::GitConfig, cred::GitCredential)
+    helpers = GitCredentialHelper[]
+
+    # https://git-scm.com/docs/gitcredentials#gitcredentials-helper
+    for entry in GitConfigIter(cfg, r"credential.*\.helper")
+        section, url, name, value = split(entry)
+        @assert name == "helper"
+
+        # Only use configuration settings where the URL applies to the git credential
+        ismatch(url, cred) || continue
+
+        # An empty credential.helper resets the list to empty
+        isempty(value) && empty!(helpers)
+
+        # Due to a bug in libgit2 iteration we may read credential helpers out of order.
+        # See: https://github.com/libgit2/libgit2/issues/4361
+        #
+        # Typically the ordering doesn't matter but does in this particular case. Disabling
+        # credential helpers avoids potential issues with using the wrong credentials or
+        # writing credentials to the wrong helper.
+        if isempty(value)
+            Base.warn_once("Resetting the helper list is currently unsupported: " *
+                 "ignoring all git credential helpers.")
+            return GitCredentialHelper[]
+        end
```

Comment:
Due to a bug with libgit2 the configuration iterator may not return results in the order in which they were written in the configuration file. For credential helpers this is only an issue when trying to use the special empty string helper which indicates that all previous helpers should be ignored.

See: https://github.com/libgit2/libgit2/issues/4361

Note that this warning does show up when running the tests.